### Getting Started with Machine Learning using Scikit-learn Python Module.


Prerequisit : Knoledge of Machine Learning : Supervised Algorithm (Regression ), Python Programming

# Introducing the Iris Dataset
Machine Learning on the Iris Dataset
1. It is a Supervised Learning Problem. To Predict the species of and Iris using the measurements
2. Its a Famous dataset to beginners.
3. We will see here Regression  Algorithm from sklearn

In [1]:
# Here we are importing SVM (Support vector Machine , Built in Iris Dataset, Numpy )
from sklearn import svm
# Import dataset to load the data
from sklearn import datasets
import seaborn as sns
import numpy as np
from sklearn.cross_validation import train_test_split

C:\Users\Shailendra\Anaconda3\lib\site-packages\sklearn\cross_validation.py:41: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.
  "This module will be removed in 0.20.", DeprecationWarning)


In [2]:
# load_iris function from datasets module
iris=datasets.load_iris()
iris1=sns.load_dataset("iris")

In [3]:
print(iris1.tail())

     sepal_length  sepal_width  petal_length  petal_width    species
145           6.7          3.0           5.2          2.3  virginica
146           6.3          2.5           5.0          1.9  virginica
147           6.5          3.0           5.2          2.0  virginica
148           6.2          3.4           5.4          2.3  virginica
149           5.9          3.0           5.1          1.8  virginica


In [4]:
# The type of dataset is saved as "bunch" object , contains Iris dataset and its attributes
print(type(iris))
print(type(iris1))

<class 'sklearn.utils.Bunch'>
<class 'pandas.core.frame.DataFrame'>


In [5]:
iris1.species.value_counts()

setosa        50
virginica     50
versicolor    50
Name: species, dtype: int64

In [6]:
# print the iris dataset
print(iris.data)

[[5.1 3.5 1.4 0.2]
 [4.9 3.  1.4 0.2]
 [4.7 3.2 1.3 0.2]
 [4.6 3.1 1.5 0.2]
 [5.  3.6 1.4 0.2]
 [5.4 3.9 1.7 0.4]
 [4.6 3.4 1.4 0.3]
 [5.  3.4 1.5 0.2]
 [4.4 2.9 1.4 0.2]
 [4.9 3.1 1.5 0.1]
 [5.4 3.7 1.5 0.2]
 [4.8 3.4 1.6 0.2]
 [4.8 3.  1.4 0.1]
 [4.3 3.  1.1 0.1]
 [5.8 4.  1.2 0.2]
 [5.7 4.4 1.5 0.4]
 [5.4 3.9 1.3 0.4]
 [5.1 3.5 1.4 0.3]
 [5.7 3.8 1.7 0.3]
 [5.1 3.8 1.5 0.3]
 [5.4 3.4 1.7 0.2]
 [5.1 3.7 1.5 0.4]
 [4.6 3.6 1.  0.2]
 [5.1 3.3 1.7 0.5]
 [4.8 3.4 1.9 0.2]
 [5.  3.  1.6 0.2]
 [5.  3.4 1.6 0.4]
 [5.2 3.5 1.5 0.2]
 [5.2 3.4 1.4 0.2]
 [4.7 3.2 1.6 0.2]
 [4.8 3.1 1.6 0.2]
 [5.4 3.4 1.5 0.4]
 [5.2 4.1 1.5 0.1]
 [5.5 4.2 1.4 0.2]
 [4.9 3.1 1.5 0.1]
 [5.  3.2 1.2 0.2]
 [5.5 3.5 1.3 0.2]
 [4.9 3.1 1.5 0.1]
 [4.4 3.  1.3 0.2]
 [5.1 3.4 1.5 0.2]
 [5.  3.5 1.3 0.3]
 [4.5 2.3 1.3 0.3]
 [4.4 3.2 1.3 0.2]
 [5.  3.5 1.6 0.6]
 [5.1 3.8 1.9 0.4]
 [4.8 3.  1.4 0.3]
 [5.1 3.8 1.6 0.2]
 [4.6 3.2 1.4 0.2]
 [5.3 3.7 1.5 0.2]
 [5.  3.3 1.4 0.2]
 [7.  3.2 4.7 1.4]
 [6.4 3.2 4.5 1.5]
 [6.9 3.1 4.

In [7]:
# retuns you total no. of datapoints are there 
iris.data.shape

(150, 4)

In [8]:
# it will give you name of the four features
iris.feature_names

['sepal length (cm)',
 'sepal width (cm)',
 'petal length (cm)',
 'petal width (cm)']

In [9]:
iris.target

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

In [10]:
# to check target values
iris.target.shape

(150,)

In [11]:
# Will give you name of flowers which are encoded as 0:Setosa, 1: versicolor, 2 : verginica
iris.target_names

array(['setosa', 'versicolor', 'virginica'], dtype='<U10')

In [12]:
# print data type of features and response
print(type(iris.data))
print(type(iris.target))

<class 'numpy.ndarray'>
<class 'numpy.ndarray'>


### Note :
#### 1. The value which we are predicting is the response, its a dependent variable.
#### 2. In Classification(Supervised Learning) , response is categorical
#### 3. In Regression  (Supervised Learning), response is continuous or ordered.

# Requirement to work with scikit-learn
1. Data and response should be seperated
2. All data should be in Numeric Form
3. Data should be in structured form (in specific shape)

In [13]:
x=iris.data
print(x.shape)

(150, 4)


In [14]:
y=iris.target
print(y.shape)

(150,)


In [15]:

x_train, x_test, y_train, y_test= train_test_split(x , y , test_size= 0.3, random_state=4)
print(x_train.shape)
print(x_test.shape)
print(y_train.shape)
print(y_test.shape)



(105, 4)
(45, 4)
(105,)
(45,)


In [16]:
#from sklearn.linear_model import LogisticRegression
#model=LogisticRegression()
model=svm.SVC(kernel='linear')
#model=svm.SVC()

In [17]:
#xtrain=x_train.reshape(-1,1)
#ytrain=y_train.reshape(-1,1)
#xtest=x_test.reshape(-1,1)
#ytest=y_test.reshape(-1,1)

In [18]:
model.fit(x_train , y_train)

SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape='ovr', degree=3, gamma='auto', kernel='linear',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False)

In [19]:
accuraccy=model.score(x_test, y_test)

In [20]:
ypred=model.predict(x_test)
print(ypred, "\nsize :", ypred.shape)
print(iris.target_names)

[2 0 2 2 2 1 2 0 0 2 0 0 0 1 2 0 1 0 0 2 0 2 1 0 0 0 0 0 0 2 1 0 2 0 1 2 2
 1 1 0 2 0 1 0 2] 
size : (45,)
['setosa' 'versicolor' 'virginica']


In [21]:
from sklearn.metrics import accuracy_score
from sklearn.metrics import classification_report

In [22]:
print(accuracy_score(y_test, ypred))

0.9777777777777777


In [23]:
print(classification_report(y_test, ypred))

             precision    recall  f1-score   support

          0       1.00      1.00      1.00        21
          1       1.00      0.90      0.95        10
          2       0.93      1.00      0.97        14

avg / total       0.98      0.98      0.98        45



In [24]:
# Now the model is train and one can test with new value
print(model.predict([[3,5,4,2]]))
print(model.predict([[6.7,3.0,5.2,2.3]]))
x_new=[[3,5,4,2],[5,4,3,2],[6.7,3.0,5.2,2.3]]
print(model.predict(x_new))


[1]
[2]
[1 1 2]


### ---------- END---------------

# Using KNN (K Nearest Neighbors) Algorithm

In [25]:

from sklearn.datasets import load_iris

iris=load_iris()
x=iris.data
y=iris.target
print(x.shape)
print(y.shape)
# KNN
#  Import the class for KNN
from sklearn.neighbors import KNeighborsClassifier

(150, 4)
(150,)


In [26]:
# create the object of KNeighborsClassifier
# to select nearest neighbors 1
knn=KNeighborsClassifier(n_neighbors=5)    # n=5
knn1=KNeighborsClassifier(n_neighbors=1)# n=1

In [27]:
print(knn)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=1, n_neighbors=5, p=2,
           weights='uniform')


#### Fit the model with data
#### Here model is learning relationship between Independent (x) variable and dependent variable(y)
1. KNN is simple Machine Learning model but it will make Highly Accurate Prediction
2. k=1, whill check nearest node and decide the class
3. k=5, will check , out of 5 how many nearest neighbors are close to the test node. 
   Then accordingly decide the test class.

In [28]:
from sklearn.cross_validation import train_test_split
x_train, x_test, y_train, y_test= train_test_split(x,y, test_size=0.3, random_state=4)

In [29]:
print(x_train.shape)
print(y_train.shape)
print(x_test.shape)

print(y_test.shape)

(105, 4)
(105,)
(45, 4)
(45,)


In [30]:
knn.fit(x_train, y_train)   # knn, n=5
knn1.fit(x_train, y_train)   # knn, n=1


KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=1, n_neighbors=1, p=2,
           weights='uniform')

In [31]:
#ypred=knn.predict(x)
#print(ypred)
ytestpred=knn.predict(x_test) # n=5
print("ytestpred = ", ytestpred)
ytestpred1=knn1.predict(x_test)  #n=1

ytestpred =  [2 0 2 2 2 1 2 0 0 2 0 0 0 1 2 0 1 0 0 2 0 2 1 0 0 0 0 0 0 2 1 0 2 0 1 2 2
 1 1 0 2 0 1 0 2]


In [32]:
from sklearn import metrics
#print(metrics.accuracy_score(y,ypred))
print("accuracy N=5", metrics.accuracy_score(y_test,ytestpred))     #knn, n=5
print("accuracy N=1",metrics.accuracy_score(y_test, ytestpred1))  # knn, n=1

accuracy N=5 0.9777777777777777
accuracy N=1 0.9555555555555556


In [33]:
# Test prediction with new value  ( This code is executed with splitting data using train test split)
print(knn.predict([[3,5,4,2]]))
x_new=[[3,5,4,2],[5,4,3,2],[6.7,3.0,5.2,2.3]]
print(knn.predict(x_new))

[1]
[1 1 2]


In [34]:
# Now the model is train and one can test with new value
print(knn.predict([[3,5,4,2]]))
print(knn.predict([[6.7,3.0,5.2,2.3]]))
x_new=[[3,5,4,2],[5,4,3,2],[6.7,3.0,5.2,2.3]]
print(knn.predict(x_new))


[1]
[2]
[1 1 2]


# Using Logistic Regression Model

In [35]:
# using Logistic Regression Model
from sklearn.linear_model import LogisticRegression
from sklearn import metrics
lgr=LogisticRegression()
lgr.fit(x,y)
lgr.predict(x)

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       2, 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 1, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

In [36]:
ypred=lgr.predict(x)

In [37]:
len(ypred)

150

In [38]:
print("Accuracy without data split : ", metrics.accuracy_score(y,ypred))

Accuracy without data split :  0.96


In [39]:
# Now the model is train and one can test with new value   (Without data split)
print(lgr.predict([[3,5,4,2]]))
print(lgr.predict([[6.7,3.0,5.2,2.3]]))
x_new=[[3,5,4,2],[5,4,3,2],[6.7,3.0,5.2,2.3]]
print(lgr.predict(x_new))


[2]
[2]
[2 0 2]


In [40]:
from sklearn.cross_validation import train_test_split
x_train, x_test, y_train, y_test= train_test_split(x,y, test_size=0.3, random_state=4)

In [41]:
# prediction after data split ,ypredtest on test data.
ytestpred=lgr.predict(x_test)
print("Accuracy after data split:", metrics.accuracy_score(y_test , ytestpred))

Accuracy after data split: 0.9555555555555556


In [42]:
# Now the model is train and one can test with new value   (With data split)
print(lgr.predict([[3,5,4,2]]))
print(lgr.predict([[6.7,3.0,5.2,2.3]]))
x_new=[[3,5,4,2],[5,4,3,2],[6.7,3.0,5.2,2.3]]
print(lgr.predict(x_new))


[2]
[2]
[2 0 2]


### Naive Bayes Algorithm

In [57]:
# import Naive Bayes Algorithm
from sklearn.naive_bayes import GaussianNB
from sklearn import metrics
from sklearn.metrics import accuracy_score
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix
model = GaussianNB()
model.fit(x_train, y_train)
ypred=model.predict(x_test)
print("Accuracy =: ", metrics.accuracy_score(y_test, ypred))

print("Classification Report := \n", classification_report(y_test, ypred))
print("Confusion matrix =:", confusion_matrix(y_test, ypred))


Accuracy =:  0.9777777777777777
Classification Report := 
              precision    recall  f1-score   support

          0       1.00      1.00      1.00        21
          1       0.91      1.00      0.95        10
          2       1.00      0.93      0.96        14

avg / total       0.98      0.98      0.98        45

Confusion matrix =: [[21  0  0]
 [ 0 10  0]
 [ 0  1 13]]


In [59]:
# Now the model is train and one can test with new value   (With data split)
print(model.predict([[3,5,4,2]]))
print(model.predict([[6.7,3.0,5.2,2.3]]))
x_new=[[3,5,4,2],[5,4,3,2],[6.7,3.0,5.2,2.3]]
print(model.predict(x_new))


[2]
[2]
[2 2 2]


### Conclusion : We have tested three Model:
                                                Linear regression using SVM : 97.77%,
                                                KNN algorithm : 97.77% (N=5) and
                                                Logistic regression : 96%
                                                Naive Bayes : 97.77%

------ END -----------